### DISCUSSION TITLE
Help with <Link>: url changes but not the page

### DISCUSSION THREAD

**Discussion by markschellhas** [vercel/next.js#18094] [Help] [answered]
I'm having an issue where my routes for a certain template page are not working at all anymore and I can't figure out why:

Profile/Contact Page is located at `/pages/contact/[id]/index.js`

If I'm on another page (say Home page) and click on a link to a contact, the contact page loads fine, for example:

**On pages/home.js**
```HTML
<Link href="/contact/1234"><a>Bill's profile page</a></Link>
```

However, on the contact page, there is a list of links to related contacts, for example:

**On pages/contact/[id]/index.js**
```HTML
<ul>
<li><Link href="/contact/1235"><a>Tom's profile page</a></Link></li>
<li><Link href="/contact/1238"><a>Jane's profile page</a></Link></li>
</ul>
```

The `pages/contact/[id]/index.js` page looks something like this:
```HTML
class Contact extends Component {

    constructor(props) {
        super(props);

        this.state = {
            contact: {},
            pageLoaded: false,
            isLoading: false,
        }

    }

    async componentDidMount() {

        const { id } = Router.query;
        await this.setState({ id });
        await this.getContact();
        this.setState({ pageLoaded: true });

    }

    getContact = () => {
        ContactService.getOne(this.state.id)
            .then(async (contact) => {
                if (!contact) {
                    Router.push("/contact/notfound")
                } else {
                    await this.setState({ contact });
                }
            })
            .catch(error => console.log(error));
    }

    render() {
        const { isLoading, contact } = this.state;

        return (
            <>
            <h1>This is {contactName}'s profile</h1>
            <h2>Related contacts</h2>
<ul>
<li><Link href="/contact/1235"><a>Tom's profile page</a></Link></li>
<li><Link href="/contact/1238"><a>Jane's profile page</a></Link></li>
</ul>

            </>
        );
    }

};
Contact.contextType = AuthContext;

export default Contact;
```

**Now the problem:** 
When I click on the link to Tom's profile page, the url in the address bar of the window changes to /contact/1235, BUT the content of the page **doesn't** change. Which means the page isn't refreshing to fetch Tom's data from the api endpoint. This used to work 100% fine, but doesn't any more. I've even updated Next.js from `9.4.4` to `9.5.4-canary.23` as I read on a SO thread, but this hasn't affected anything.

Can someone please help me out here? I'm severely confused and this is breaking a critical function in my app. 🤯

**Comment by jamesmosier** [comment, marked as answer]
Hi @markschellhas. I believe the reason is because your component does not know it needs to update. When you click the link to go to a new contact, `componentDidMount` does not run again. Therefore you need to "watch" for changes and fetch the new data and update it in state.

To resolve this, you could use the [`withRouter`](https://nextjs.org/docs/api-reference/next/router#withrouter) higher order component and something like:

```js
async componentDidUpdate(prevProps) {
  const { query } = this.props.router
  // verify props have changed to avoid an infinite loop
  if (query.id !== prevProps.router.query.id) {
    // fetch data based on the new query
    this.setState({ id: query.id });
    await this.getContact(query.id);
  }
}
```

and you can change `getContact` to accept an id:

```js
getContact = (id) => {
```

because `setState` is async, therefore `this.state.id` may not be updated by the time you call `getContact`. So by passing in a parameter you can be sure it is up-to-date.

**Reply by markschellhas** [reply]
@jamesmosier Of course, makes total sense, thanks for helping me out of a rut! 🙏🏼
